src/hidra/receiver/asapo_transfer.py

```diff
@@ -0,0 +1,208 @@
+            [metadata, data] = self.query.get(timeout=self.reconnect_timeout*1000)
+            if metadata is not None:
+                try:
+                    local_path = generate_filepath(self.target_dir, metadata)
+                    logger.info("Send file {local_path}".format(local_path=local_path))
+                    self.asapo_worker.send_message(local_path, metadata)
+                except Exception as e:
+                    logger.error("Transmission does not succeed. {e}. File ignored".format(e=str(e)))
+
+    def stop(self):
+        logger.info("Runner is stopped.")
+        self.stop_run.set()
+
+
+def main():
+    parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter,
+                                     description='Transfer files metadata to ASAPO')
+    parser.add_argument('--config_path', type=str, help='Path to config directory')
+    parser.add_argument(
+        'identifier', type=str,
+        help='Beamline and detector ID information separated by a single underscore')
+
+    args = vars(parser.parse_args())
+    config = construct_config(args.pop('config_path'), args.pop('identifier'))
+    logging.basicConfig(format="%(asctime)s %(module)s %(lineno)-6d %(levelname)-6s %(message)s",
+                        level=getattr(logging, config.log_level))
+    logger.info("Start Asapo transfer with parameters %s", config)
+
+    worker_args = dict(
+        endpoint=config.endpoint,
+        beamtime=config.beamtime,
+        token=config.token,
+        n_threads=config.n_threads,
+        file_regex=config.file_regex,
+        default_data_source=config.default_data_source,
+        timeout=config.timeout,
+        beamline=config.beamline,
+        start_file_idx=config.start_file_idx)
+
+    logger.info("Creating AsapoWorker with %s", worker_args)
+    asapo_worker = AsapoWorker(**worker_args)
+    run_transfer(asapo_worker, config, config.reconnect_timeout)
+
+
+def construct_config(config_path, identifier):
+
+    # Read config file
+    with open(f"{config_path}/asapo_transfer_{identifier}.yaml", "r") as f:
+        config = yaml.load(f.read(), Loader=yaml.FullLoader)
+
+    # parse identifier to extruct `beamline` and `detector_id`
+    beamline, detector_id = identifier.split("_", maxsplit=1)
+    config['beamline'] = beamline
+    config['detector_id'] = detector_id
+
+    if 'default_data_source' not in config:
+        config['default_data_source'] = detector_id
+        # remove `.desy.de` if present at the end
+        if detector_id[-8:] == ".desy.de":
+            config['default_data_source'] = detector_id[:-8]
+
+    config['target_host'] = socket.getfqdn()
+    config['signal_host'] = _constants.CONNECTION_LIST[beamline]['host']
+
+    return TransferConfig(**config)
+
+
+def run_transfer(asapo_worker, config, timeout=3):
+
+    query = create_query(config.signal_host, config.detector_id)
+    asapo_transfer = create_asapo_transfer(asapo_worker, query,
```

Comment:
You could move these into the while (on probably also try) block. Then you don't need to repeat them later (it could be better to recreate everything for every error, who knows the state of `query` after an arbitrary error occurred).

Reply:
Does it make sense to use new port on each error?

Reply:
If yes, there is no reason to catch zmq error separately.

Reply:
Now AsapoTransfer object is created on each error.